sdk/pinocchio/src/sysvars/slot_hashes.rs

```diff
@@ -0,0 +1,895 @@
+//! Efficient, zero-copy access to SlotHashes sysvar data.
+
+use crate::{
+    account_info::{AccountInfo, Ref},
+    program_error::ProgramError,
+    pubkey::Pubkey,
+    sysvars::clock::Slot,
+};
+use core::{mem, ops::Deref, ptr};
+
+/// SysvarS1otHashes111111111111111111111111111
+pub const SLOTHASHES_ID: Pubkey = [
+    6, 167, 213, 23, 25, 47, 10, 175, 198, 242, 101, 227, 251, 119, 204, 122, 218, 130, 197, 41,
+    208, 190, 59, 19, 110, 45, 0, 85, 32, 0, 0, 0,
+];
+pub const MAX_ENTRIES: usize = 512;
+pub const HASH_BYTES: usize = 32;
+
+/// Sysvar data is:
+/// len    (8 bytes): little-endian entry count (≤ 512)
+/// entries(len × 40 bytes):    consecutive `(u64 slot, [u8;32] hash)` pairs
+pub const NUM_ENTRIES_SIZE: usize = mem::size_of::<u64>();
+pub const SLOT_SIZE: usize = mem::size_of::<Slot>();
+pub const ENTRY_SIZE: usize = SLOT_SIZE + HASH_BYTES;
+
+/// A single entry in the `SlotHashes` sysvar.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+#[repr(C)]
+pub struct SlotHashEntry {
+    /// The slot number.
+    pub slot: Slot,
+    /// The hash corresponding to the slot.
+    pub hash: [u8; HASH_BYTES],
+}
+
+/// Provides zero-copy access to the data of the `SlotHashes` sysvar.
+///
+/// Internally it keeps either a plain slice `&'a [u8]` *or* the `Ref<'a, [u8]>`
+/// returned by `AccountInfo::try_borrow_data()`.  Holding the `Ref` variant is
+/// important because dropping the `Ref` would release the runtime borrow while
+/// users still hold `&[u8]` references obtained from the struct.
+enum Data<'a> {
+    Slice(&'a [u8]),
+    Ref(Ref<'a, [u8]>),
+}
+
+impl core::ops::Deref for Data<'_> {
+    type Target = [u8];
+    #[inline(always)]
+    fn deref(&self) -> &Self::Target {
+        match self {
+            Data::Slice(s) => s,
+            Data::Ref(r) => r,
+        }
+    }
+}
+
+/// SlotHashes provides read-only, zero-copy access to SlotHashes sysvar bytes.
+pub struct SlotHashes<'a> {
+    data: Data<'a>,
+    len: usize,
+}
+
+impl<'a> SlotHashes<'a> {
+    /// Creates a `SlotHashes` instance directly from a data container and entry count.
+    /// Important: provide a valid len. Whether or not len is assumed to be
+    /// the constant 20_488 (512 entries) is up to caller.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe because it does not check the validity of the data or count.
+    /// The caller must ensure:
+    /// 1. The underlying byte slice in `data` represents valid SlotHashes data
+    ///    (length prefix + entries).
+    /// 2. `len` is the correct number of entries (≤ MAX_ENTRIES), matching the prefix.
+    /// 3. The data slice contains at least `NUM_ENTRIES_SIZE + len * ENTRY_SIZE` bytes.
+    /// 4. If `T` is `&[u8]`, that borrow rules are upheld.
+    /// 5. Alignment is correct.
+    ///
+    #[inline(always)]
+    pub unsafe fn new_unchecked_slice(data: &'a [u8], len: usize) -> Self {
```

Comment:
If a generic `T: Deref<Target = [u8]>` is used, then we only need one `new_unchecked` constructor.

Reply:
Indeed, addressed in dce5e60da08b583d82c3eee70c6114ba52a2bc8b and 82336188de4e446676dac8b8cd078bd24f2aa646